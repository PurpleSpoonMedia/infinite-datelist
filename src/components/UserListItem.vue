<template>
  <tr>
    <td>{{ user.name }}</td>
    <td>{{ user.email }}</td>
    <td>{{ user.phone }}</td>
    <td>{{ user.country }}</td>
    <td>{{ user.services }}</td>
    <td class="text-right">
      <button class="btn btn-info btn-sm" @click="showDetailsModal = true">Details</button>
      <router-link class="btn btn-success btn-sm" :to="`/edituser/${user.id}`">Edit</router-link>
      <button class="btn btn-danger btn-sm" @click="showDeleteModal = true">Delete</button>
    </td>
    <!-- Modal Delete -->
    <app-modal v-if="showDeleteModal" >
      <h5 slot="header" class="modal-title">
        Delete user
      </h5>
      <p slot="body">
        Are you sure you want to delete <b>{{ user.name }}</b>&apos;s profile?
      </p>
      <div slot="footer">
        <button
          type="button"
          class="btn btn-danger"
          @click="handleDeleteUser"
        >
          Delete
        </button>
        <button
          type="button"
          class="btn btn-secondary"
          data-dismiss="modal"
          @click="showDeleteModal = false"
        >
          Cancel
        </button>
      </div>
    </app-modal>
    <!-- /Modal Delete -->
    <!-- Modal Details -->
    <app-modal v-if="showDetailsModal" >
      <h5 slot="header" class="modal-title">
        <b>{{ user.name }}</b>
      </h5>
      <p slot="body">
        <ul class="list-unstyled">
          <li>Name: {{ user.name }}</li>
          <li>Email: {{ user.email }}</li>
          <li>Phone: {{ user.phone }}</li>
          <li>Country: {{ user.country }}</li>
          <li>Services: {{ user.services }}</li>
          <li>Newsletter: {{ user.newsletter }}</li>
        </ul>
      </p>
      <div slot="footer">
        <button
          type="button"
          class="btn btn-secondary"
          data-dismiss="modal"
          @click="showDetailsModal = false"
        >
          Cancel
        </button>
      </div>
    </app-modal>
    <!-- /Modal Details -->
  </tr>
</template>

<script>
import Modal from './Modal';

export default {
  props: {
    user: Object
  },
  data() {
    return {
      showDeleteModal: false,
      showDetailsModal: false,
    }
  },
  methods: {
    handleDeleteUser() {
      this.$store.dispatch('deleteUser', this.user.id);
    }
  },
  components: {
    appModal: Modal
  }
}
</script>
