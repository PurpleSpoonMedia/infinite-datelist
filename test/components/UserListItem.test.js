import { mount, RouterLinkStub } from '@vue/test-utils';
import UserListItem from '../../src/components/UserListItem.vue';
import users from '../fixtures/users';

describe('UserListItem.vue', () => {
  let wrapper, handleDeleteUser;
  beforeEach(() => {
    handleDeleteUser = jest.fn();
    wrapper = mount(UserListItem, {
      propsData: {
        user: users[0]
      },
      stubs: {
        RouterLink: RouterLinkStub
      },
      methods: {
        handleDeleteUser
      }
    })
  });

  test('should render UserListItem correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should open modal when click on Details button', () => {
    const detailsButton = wrapper.find('#details');
    detailsButton.trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should open confirmation modal when click on Delete button then call handleDeleteUser method when click on Delete button in modal', () => {
    const deleteButton = wrapper.find('#delete-modal');
    deleteButton.trigger('click');
    expect(wrapper.html()).toMatchSnapshot();

    const deleteModal = wrapper.find('#delete');
    deleteModal.trigger('click');
    expect(handleDeleteUser).toHaveBeenCalled();
  });
});
