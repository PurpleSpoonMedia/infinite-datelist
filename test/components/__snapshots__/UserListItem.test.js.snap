// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListItem.vue should open confirmation modal when click on Delete button then call handleDeleteUser method when click on Delete button in modal 1`] = `
<tr>
  <td>Test 1</td>
  <td>test1@example.com</td>
  <td>123456</td>
  <td>Hungary</td>
  <td>mobile, internet</td>
  <td class="text-right">
    <button id="details" class="btn btn-info btn-sm">Details</button>
    <a class="btn btn-success btn-sm">Edit</a>
    <button id="delete-modal" class="btn btn-danger btn-sm">Delete</button>
  </td>
  <div class="modal modal-mask" style="display: block;" name="fade">
    <div role="document" class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title">
            Delete user
          </h5>
        </div>
        <div class="modal-body">
          <p>
            Are you sure you want to delete
            <b>Test 1</b>'s profile?
          </p>
        </div>
        <div class="modal-footer">
          <div>
            <button type="button" id="delete" class="btn btn-danger">
              Delete
            </button>
            <button type="button" data-dismiss="modal" class="btn btn-secondary">
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <!---->
</tr>
`;

exports[`UserListItem.vue should open modal when click on Details button 1`] = `
<tr>
  <td>Test 1</td>
  <td>test1@example.com</td>
  <td>123456</td>
  <td>Hungary</td>
  <td>mobile, internet</td>
  <td class="text-right">
    <button id="details" class="btn btn-info btn-sm">Details</button>
    <a class="btn btn-success btn-sm">Edit</a>
    <button id="delete-modal" class="btn btn-danger btn-sm">Delete</button>
  </td>
  <!---->
  <div class="modal modal-mask" style="display: block;" name="fade">
    <div role="document" class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title">
            <b>Test 1</b>
          </h5>
        </div>
        <div class="modal-body">
          <p>
            <ul class="list-unstyled">
              <li>Name: Test 1</li>
              <li>Email: test1@example.com</li>
              <li>Phone: 123456</li>
              <li>Country: Hungary</li>
              <li>Services: mobile, internet</li>
              <li>Newsletter: true</li>
            </ul>
          </p>
        </div>
        <div class="modal-footer">
          <div>
            <button type="button" data-dismiss="modal" class="btn btn-secondary">
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</tr>
`;

exports[`UserListItem.vue should render UserListItem correctly 1`] = `
<tr>
  <td>Test 1</td>
  <td>test1@example.com</td>
  <td>123456</td>
  <td>Hungary</td>
  <td>mobile, internet</td>
  <td class="text-right">
    <button id="details" class="btn btn-info btn-sm">Details</button>
    <a class="btn btn-success btn-sm">Edit</a>
    <button id="delete-modal" class="btn btn-danger btn-sm">Delete</button>
  </td>
  <!---->
  <!---->
</tr>
`;
